File: src/main/java/dev/goerner/geozen/model/MultiLineString.java
package dev.goerner.geozen.model;

import java.util.ArrayList;

/**
 * A {@link MultiLineString} is a {@link Geometry} that represents a collection of {@link LineString LineStrings} in
 * space. It is defined by a list of {@link LineString LineStrings} and a {@link CoordinateReferenceSystem}.
 */
public class MultiLineString extends Geometry {

	private ArrayList<ArrayList<Position>> coordinates;

	/**
	 * Creates a new {@link MultiLineString} with an empty list of {@link LineString LineStrings} and the default WGS 84
	 * {@link CoordinateReferenceSystem}.
	 */
	public MultiLineString() {
		this(new ArrayList<>());
	}

	/**
	 * Creates a new {@link MultiLineString} with the given list of {@link LineString LineStrings} and the default WGS 84
	 * {@link CoordinateReferenceSystem}.
	 *
	 * @param coordinates The list of {@link LineString LineStrings} representing the {@link MultiLineString}.
	 */
	public MultiLineString(ArrayList<ArrayList<Position>> coordinates) {
		this(coordinates, CoordinateReferenceSystem.WGS_84);
	}

	/**
	 * Creates a new {@link MultiLineString} with the given list of {@link LineString LineStrings} and the given
	 * {@link CoordinateReferenceSystem}.
	 *
	 * @param coordinates               The list of {@link LineString LineStrings} representing the {@link MultiLineString}.
	 * @param coordinateReferenceSystem The {@link CoordinateReferenceSystem} of the {@link MultiLineString}.
	 */
	public MultiLineString(ArrayList<ArrayList<Position>> coordinates, CoordinateReferenceSystem coordinateReferenceSystem) {
		super(coordinateReferenceSystem);
		this.coordinates = coordinates;
	}

	@Override
	public double getMinDistanceTo(Point point, boolean exact) {
		return 0;
	}

	@Override
	public double getMinDistanceTo(LineString lineString, boolean exact) {
		return 0;
	}

	@Override
	public double getMinDistanceTo(Polygon polygon, boolean exact) {
		return 0;
	}

	@Override
	public double getMinDistanceTo(MultiPoint multiPoint, boolean exact) {
		return 0;
	}

	@Override
	public double getMinDistanceTo(MultiLineString multiLineString, boolean exact) {
		return 0;
	}

	@Override
	public double getMinDistanceTo(MultiPolygon multiPolygon, boolean exact) {
		return 0;
	}

	public void addLineString(ArrayList<Position> lineString) {
		this.coordinates.add(lineString);
	}

	public void addLineString(LineString lineString) {
		this.coordinates.add(lineString.getCoordinates());
	}

	public ArrayList<ArrayList<Position>> getCoordinates() {
		return coordinates;
	}

	public void setCoordinates(ArrayList<ArrayList<Position>> coordinates) {
		this.coordinates = coordinates;
	}
}

File: src/main/kotlin/dev/goerner/geozen/model/multi_geometry/MultiLineString.kt
package dev.goerner.geozen.model.multi_geometry

import dev.goerner.geozen.calc.ApproximateDistanceCalculator
import dev.goerner.geozen.calc.PreciseDistanceCalculator
import dev.goerner.geozen.model.CoordinateReferenceSystem
import dev.goerner.geozen.model.Geometry
import dev.goerner.geozen.model.Position
import dev.goerner.geozen.model.Reprojector

/**
 * A [MultiLineString] is a [Geometry] that represents a collection of
 * [LineStrings][dev.goerner.geozen.model.simple_geometry.LineString] in space. It is defined by a
 * list of [LineStrings][dev.goerner.geozen.model.simple_geometry.LineString] and a
 * [CoordinateReferenceSystem].
 *
 * @param coordinates The list of line strings (each represented as a list of positions). The outer
 *   list may be empty, but each inner list must contain at least 2 positions.
 * @param coordinateReferenceSystem The coordinate reference system, defaults to WGS_84
 * @throws IllegalArgumentException if any inner list contains fewer than 2 positions
 */
data class MultiLineString
@JvmOverloads
constructor(
    val coordinates: List<List<Position>>,
    override val coordinateReferenceSystem: CoordinateReferenceSystem =
        CoordinateReferenceSystem.WGS_84,
) : Geometry(coordinateReferenceSystem) {

    init {
        coordinates.forEachIndexed { index, lineString ->
            require(lineString.size >= 2) {
                "Each LineString in MultiLineString must contain at least 2 positions, " +
                    "but LineString at index $index contained ${lineString.size}"
            }
        }
    }

    override fun fastDistanceTo(other: Geometry): Double =
        ApproximateDistanceCalculator.calculate(this, other)

    override fun exactDistanceTo(other: Geometry): Double =
        PreciseDistanceCalculator.calculate(this, other)

    override fun reprojectTo(crs: CoordinateReferenceSystem): Geometry =
        if (coordinateReferenceSystem == crs) this
        else
            MultiLineString(
                coordinates.map { lineString ->
                    lineString.map { Reprojector.reproject(it, coordinateReferenceSystem, crs) }
                },
                crs,
            )
}
